### PR TITLE
Add `invoke` command

### LINKED ISSUE #883: Add `invoke` command

### PR DESCRIPTION
**Pull request by kasperski95**
Closes #883

### CODE REVIEW

File: protostar/commands/invoke_command.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            *NetworkCommandUtil.network_arguments,
+            ProtostarArgument(
+                name="contract-address",
+                description="The address of the contract being called.",
+                type="int",
+                is_required=True,
+            ),
+            ProtostarArgument(
+                name="function",
+                description="The name of the function being called.",
```
Comment (Arcticae) [inline protostar/commands/invoke_command.py:52]:
```suggestion
                description="The name of the function from contract's interface.",
```

File: protostar/commands/invoke_command.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                type="str",
+                is_required=True,
+            ),
+            ProtostarArgument(
+                name="inputs",
+                description="The name of the function being called.",
```
Comment (Arcticae) [inline protostar/commands/invoke_command.py:58]: Description is wrong

File: protostar/commands/invoke_command.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                type="felt",
```
Comment (Arcticae) [inline protostar/commands/invoke_command.py:59]: would it be possible to have data-transformer work here? as in passing the args as json maybe. not in this PR though

File: protostar/commands/invoke_command.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        address: int,
+        function: str,
+        inputs: list[int],
+        gateway_client: GatewayClient,
+        signer: Optional[BaseSigner] = None,
+        account_address: Optional[str] = None,
+        max_fee: Optional[Fee] = None,
+        wait_for_acceptance: bool = False,
+    ):
+        gateway_facade = self._gateway_facade_factory.create(
```
Comment (Arcticae) [inline protostar/commands/invoke_command.py:108]: is logger missing here on purpose, to silence the output?

File: protostar/commands/invoke_command.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                "Argument `account_address` is required for transactions V1."
+            )
+        if max_fee is None:
+            raise ProtostarException(
+                "Argument `max-fee` is required for transactions V1."
+            )
+        if signer is None:
+            raise ProtostarException(
+                "Argument `signer` is required for transactions V1 when private-key is not detected."
+            )
```
Comment (Arcticae) [inline protostar/commands/invoke_command.py:122]: [nitpick]: This logic was implemented somewhere already - it'd be nice to have it extracted into common part.

File: protostar/commands/invoke_command.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            signer=signer,
+            max_fee=args.max_fee,
+            wait_for_acceptance=args.wait_for_acceptance,
+            account_address=args.account_address,
+        )
+
+    async def invoke(
+        self,
+        address: int,
+        function: str,
```
Comment (Arcticae) [inline protostar/commands/invoke_command.py:100]: [nitpick]: those names are ambiguous - `contract_address` and `function_name` are a bit longer but do clarify code significantly

File: protostar/commands/invoke_command.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            contract_address=address,
+            function_name=function,
+            inputs=inputs,
+            max_fee=max_fee if isinstance(max_fee, int) else None,
+            auto_estimate_fee=max_fee == "auto",
+            signer=signer,
+            account_address=account_address,
+            wait_for_acceptance=wait_for_acceptance,
+        )
+        self._logger.info(self.format_successful_invoke_response(response))
```
Comment (Arcticae) [inline protostar/commands/invoke_command.py:134]: what about failed result? we need to print out some info as well there

File: tests/integration/protostar_fixture.py [hunk 8]
```diff
@@ -3,50 +3,47 @@
...
+from protostar.migrator import Migrator, MigratorExecutionEnvironment
+from protostar.protostar_toml import ProtostarContractsSection, ProtostarTOMLWriter
+from protostar.protostar_toml.io.protostar_toml_reader import ProtostarTOMLReader
+from protostar.protostar_toml.protostar_project_section import ProtostarProjectSection
+from protostar.self.protostar_directory import ProtostarDirectory
+from protostar.starknet_gateway import Fee, GatewayFacadeFactory
+from protostar.testing import TestingSummary
 
 
+# pylint: disable=too-many-instance-attributes
```
Comment (Arcticae) [inline tests/integration/protostar_fixture.py:46]: how about removing this lint? we're disabling it all over the place 😬

**Review by Arcticae — CHANGES_REQUESTED** [review]
We're missing some tests for providing a custom signer from using the CLI

File: protostar/commands/invoke_command.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        return None
+
+    @property
+    def arguments(self):
+        return [
+            *SignableCommandUtil.signable_arguments,
+            *NetworkCommandUtil.network_arguments,
+            ProtostarArgument(
+                name="contract-address",
+                description="The address of the contract being called.",
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/invoke_command.py:46]: https://docs.starknet.io/documentation/develop/CLI/commands/#starknet_invoke

File: protostar/commands/invoke_command.py (hunk 1 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py:52]: https://docs.starknet.io/documentation/develop/CLI/commands/#starknet_invoke

File: protostar/commands/invoke_command.py (hunk 5 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py:122]: We can make signing args required when we drop support for V0.

File: protostar/commands/invoke_command.py (hunk 3 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py:59]: It's possible, but this is out of scope of this PR.
https://github.com/software-mansion/protostar/issues/992

File: protostar/commands/invoke_command.py (hunk 4 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py:108]: On purpose. Logger is used by migrations to show data flowing in and out. We don't do that for declare and deploy commands.

File: protostar/commands/invoke_command.py (hunk 6 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py:100]: I tried to be consistent with `starknet invoke`. I'll make names more explicit.

https://docs.starknet.io/documentation/develop/CLI/commands/#starknet_invoke

File: protostar/commands/invoke_command.py (hunk 1 above)
Comment (Arcticae) [reply protostar/commands/invoke_command.py:52]: i understand, but if we can do better - we should :)

File: website/docs/tutorials/06-deploying/01-cli.md [hunk 10]
```diff
@@ -61,6 +61,11 @@ Protostar synchronizes `PYTHONPATH` with project's `cairo_path`.
 Modules that are dependencies of Protostar (like `starknet_py` or `cairo-lang`) should be available for importing by default.
 If you want to import other custom modules, you should extend `PYTHONPATH` yourself (https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH), when running this command.
 
+## Sending Invoke Transaction
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply website/docs/tutorials/06-deploying/01-cli.md:64]: https://github.com/software-mansion/protostar/issues/993

I don't know what should I write there tbh. I think CLI docs should be good enough. StarkNet describes invoke in [StarkNet CLI reference](https://docs.starknet.io/documentation/develop/CLI/commands/).

File: tests/integration/protostar_fixture.py (hunk 8 above)
Comment (kasperski95) [reply tests/integration/protostar_fixture.py:46]: I am not sure. I found 4 similar comments in the code base.

File: protostar/commands/invoke_command.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+            raise ProtostarException(
+                "Argument `signer` is required for transactions V1 when private-key is not detected."
+            )
+
+        response = await gateway_facade.invoke(
+            contract_address=address,
+            function_name=function,
+            inputs=inputs,
+            max_fee=max_fee if isinstance(max_fee, int) else None,
+            auto_estimate_fee=max_fee == "auto",
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/invoke_command.py:129]: It's more a scope of this PR https://github.com/software-mansion/protostar/pull/979.

File: protostar/commands/invoke_command.py (hunk 1 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py:52]: @view defines contract interface yet (I think) can't be invoked

**Review by mkaput — APPROVED** [review]

File: tests/integration/protostar_fixture.py (hunk 8 above)
Comment (Arcticae) [reply tests/integration/protostar_fixture.py:46]: yeah exactly, so why do we need it?

Comment (Arcticae) [reply tests/integration/protostar_fixture.py:46]: if it breaks the rule, we ignore it. so maybe it's better to disable it overall

File: protostar/cli/network_command_util.py [hunk 12]
```diff
@@ -91,4 +91,13 @@ def normalize_chain_id(
...
+            supported_chain_ids = [
+                f"- {chain_id.value} ({chain_id.name})" for chain_id in StarknetChainId
+            ]
+            formatted_chain_ids = "\n".join(supported_chain_ids)
+
+            raise ProtostarException(
+                "Invalid chain id value.\n"
+                "Supported chain ids:\n"
+                f"{formatted_chain_ids}"
+            ) from ex
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/cli/network_command_util.py:103]: It is, but super-small PRs generates extra overhead.

File: tests/integration/protostar_fixture.py (hunk 8 above)
Comment (kasperski95) [reply tests/integration/protostar_fixture.py:48]: This rule encourages rethinking design. Anyway, disabling this rule is out of scope of this PR.

File: tests/integration/commands/invoke_command_test.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@pytest.fixture(name="protostar")
+def protostar_fixture(create_protostar_project: CreateProtostarProjectFixture):
+    with create_protostar_project() as protostar:
+        protostar.build_sync()
+        yield protostar
+
+
+async def test_invoke(
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply tests/integration/commands/invoke_command_test.py:18]: I added extra asserts to check if values are actually respected. We need to implemented `CallCommand` to create a more stable test. I think parametrizing isn't necessary in this case.

File: protostar/commands/invoke_command.py (hunk 7 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py:133]: Added a test case. `GatewayFacade` raises `ProtostarException`.

File: protostar/commands/invoke_command.py (hunk 2 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py]: Fixed

File: protostar/commands/invoke_command.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/invoke_command.py:46]: I left the original description for this argument.

**Comment by kasperski95** [comment]
> We're missing some tests for providing a custom signer from using the CLI

I'd say it's out of scope of this PR. https://github.com/software-mansion/protostar/issues/879

File: protostar/cli/network_command_util.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mkaput) [reply protostar/cli/network_command_util.py:103]: I do not agree, especially given this snippet is irrelevant to PRs scope

[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/cli/network_command_util.py:103]: It's a related issue, I needed to manually specify chain-id when testing the invoke command.

Ok, I'll remove this change from this PR. IMO, creating super-focused PRs discourage developers from adding small improvements.

**Review by Arcticae — APPROVED** [review]

File: protostar/commands/invoke_command.py (hunk 6 above)
Comment (kasperski95) [reply protostar/commands/invoke_command.py]: Renamed. Resolving.

**Pull request merged**